### DISCUSSION TITLE
Overload resolution with integer parameters for default literal

### DISCUSSION THREAD

**Discussion by ufcpp** [dotnet/csharplang#7384] [Q&A] [answered]
Recently, I found the following behavior:

```cs
X.M(0);       // M(int) 
X.M(default); // M(byte)

X.N(0);       // N(byte?)
X.N(default); // N(byte?)
X.N(null);    // N(byte?)

class X
{
    public static void M(int i32) { }
    public static void M(byte i8) { }

    public static void N(int? i32) { }
    public static void N(byte? i8) { }
}
```

The question is:

* Why is `N(0)` resolved as `N(byte)`? Accoding to the [spec](https://github.com/dotnet/csharpstandard/blob/b90146145ea396924aab57d1e73e4246ea8ad359/standard/lexical-structure.md), `0` is `int`.
* Do you think it's OK that `M(0)` and `M(default)` behave differently?
* Is there any spec for `M(default)` is resolved as `M(byte)` and `N(null)` is `N(byte?)`?

**Comment by khm1600** [comment, marked as answer]
If my understanding of the spec is correct: 

`0` is `int`, therefore `M(0)` calls `M(int)`.

`default` is not `int` or `byte`, `int?` or `byte?`, `0` and `null` are not `int?` or `byte?`, but all of them can be implicitly converted to the types, so they are resolved just like any other case where the expression converts to both types: the function whose parameter type implicitly converts to the parameter type of the other function is better.

spec:
>6.4.5.3 Integer literals
The type of an integer literal is determined as follows:
> * If the literal has no suffix, it has the first of these types in which its value can be represented: int, uint, long, ulong.

>6.4.5.7 The null literal
A null_literal represents a null value. It does not have a type, but can be converted to any reference type or nullable value type through a null literal conversion (§10.2.7).

> 12.8.20 Default value expressions
A default_literal represents a default value (§9.3). It does not have a type, but can be converted to any type through a default literal conversion (§10.2.16).

> 12.6.4.3 Better function member
Given an argument list A with a set of argument expressions {E₁, E₂, ..., Eᵥ} and two applicable function members Mᵥ and Mₓ with parameter types {P₁, P₂, ..., Pᵥ} and {Q₁, Q₂, ..., Qᵥ}, Mᵥ is defined to be a better function member than Mₓ if
> * for each argument, the implicit conversion from Eᵥ to Qᵥ is not better than the implicit conversion from Eᵥ to Pᵥ, and
> * for at least one argument, the conversion from Eᵥ to Pᵥ is better than the conversion from Eᵥ to Qᵥ.

>12.6.4.5 Better conversion from expression
Given an implicit conversion C₁ that converts from an expression E to a type T₁, and an implicit conversion C₂ that converts from an expression E to a type T₂, C₁ is a better conversion than C₂ if one of the following holds:
> * E exactly matches T₁ and E does not exactly match T₂ (§12.6.4.6))
> * E exactly matches both or neither of T₁ and T₂, and T₁ is a better conversion target than T₂ (§12.6.4.7))

>12.6.4.7 Better conversion target
Given two types T₁ and T₂, T₁ is a better conversion target than T₂ if one of the following holds:
> * An implicit conversion from T₁ to T₂ exists and no implicit conversion from T₂ to T₁ exists

**Reply by ufcpp** [reply]
Conversion to byte is better than to int? Smaller strage is better to covert?

**Reply by CyrusNajmabadi** [reply]
yes.  consider something like 

```c#
void Foo(Wide);
void Foo(Narrow);
Foo(narrowest);
```

Which would you expect that to go to?  For example, say you had things like `string` vs `object`.  Generally people would expect to call into the narrower of the two given both being applicable.

**Comment by ufcpp** [comment]
OK, now I understood the following sentence inplies that narrower is better.

> An implicit conversion from T₁ to T₂ exists and no implicit conversion from T₂ to T₁ exists
